server/internal/functions/process.go

```diff
@@ -0,0 +1,300 @@
+	}
+	entrypointKeys := slices.Sorted(maps.Keys(validEntrypoint))
+
+	foundEntrypoint := false
+	var manifestFile *zip.File
+	for _, f := range rdr.File {
+		if _, ok := validEntrypoint[f.Name]; ok {
+			foundEntrypoint = true
+			continue
+		}
+		if f.Name == "manifest.json" {
+			manifestFile = f
+		}
+	}
+
+	if !foundEntrypoint {
+		return nil, oops.E(oops.CodeBadRequest, errPermanent, "%s: functions zip file is missing entrypoint file: %v", entrypointKeys, slug).Log(ctx, logger, attrError)
+	}
+
+	if manifestFile == nil {
+		return nil, oops.E(oops.CodeBadRequest, errPermanent, "%s: functions zip file is missing manifest.json file", slug).Log(ctx, logger, attrError)
+	}
+
+	manifestRdr, err := manifestFile.Open()
+	if err != nil {
+		return nil, oops.E(oops.CodeBadRequest, err, "%s: error opening manifest file", slug).Log(ctx, logger, attrError)
+	}
+	defer o11y.LogDefer(ctx, logger, func() error {
+		return manifestRdr.Close()
+	})
+
+	manifestBs, err := io.ReadAll(manifestRdr)
+	if err != nil {
+		return nil, oops.E(oops.CodeBadRequest, err, "%s: error reading manifest file", slug).Log(ctx, logger, attrError)
+	}
+
+	var manifest Manifest
+	if err := manifest.UnmarshalJSON(manifestBs); err != nil {
+		return nil, oops.E(oops.CodeBadRequest, err, "%s: error parsing manifest file", slug).Log(ctx, logger, attrError)
+	}
+
+	dbtx, err := p.db.Begin(ctx)
+	if err != nil {
+		return nil, oops.E(oops.CodeUnexpected, err, "%s: error opening database transaction", slug).Log(ctx, p.logger, attrError)
+	}
+	defer o11y.NoLogDefer(func() error {
+		return dbtx.Rollback(ctx)
+	})
+
+	tx := repo.New(dbtx)
+
+	numTools := 0
+	for idx, tool := range manifest.V0.Tools {
+		_, err := processManifestToolV0(
+			ctx,
+			tx,
+			tool,
+			deploymentID,
+			attachementID,
+			slug,
+			attachement.Runtime,
+		)
+		if err != nil {
+			msg := fmt.Sprintf("%s: skipping tool %d (%q): %v", slug, idx, tool.Name, err)
+			logger.ErrorContext(ctx, msg, attrError)
+			if task.OnToolSkipped != nil {
+				task.OnToolSkipped(err)
+			}
+			continue
+		}
+
+		numTools += 1
+	}
+
+	if err := dbtx.Commit(ctx); err != nil {
+		return nil, oops.E(oops.CodeUnexpected, oops.Permanent(err), "%s: error saving tools", slug).Log(ctx, logger)
+	}
+
+	return &ToolExtractorResult{
+		ManifestVersion: manifest.Version,
+		NumTools:        numTools,
+	}, nil
+}
+
+func processManifestToolV0(
+	ctx context.Context,
+	tx *repo.Queries,
+	tool ManifestToolV0,
+	deploymentID uuid.UUID,
+	attachementID uuid.UUID,
+	attachementSlug types.Slug,
+	runtime string,
+) (*repo.FunctionToolDefinition, error) {
+	if err := validateManifestToolV0(tool); err != nil {
+		return nil, tagError("invalid-manifest", "validate tool: %w", err)
+	}
+
+	name := tool.Name
+	description := tool.Description
+	inputSchema := tool.InputSchema
+	variables := tool.Variables
+
+	if len(inputSchema) == 0 {
+		inputSchema = json.RawMessage(`{}`)
```

Comment:
Should we instead leave this empty? That's what we do right now for http_tools if there is no input schema.